src/serializer.js

```diff
@@ -91,8 +91,86 @@ class FullPacketParser extends Transform {
   }
 }
 
+function recursiveGet (target, path, receiver) {
+  return path.reduce((object, key) => Reflect.get(object, key, receiver), target)
+}
+
+class LazyPacketParser extends Transform {
+  constructor (proto, mainType, shallowType, noErrorLogging = false) {
+    super({ readableObjectMode: true })
+    this.proto = proto
+    this.mainType = mainType
+    this.shallowType = shallowType
+    this.noErrorLogging = noErrorLogging
+  }
+
+  parsePacketBuffer (buffer) {
+    let shallowPacketData
+    try {
+      shallowPacketData = this.proto.parsePacketBuffer(this.shallowType, buffer).data
+    } catch (e) {
+      return this.proto.parsePacketBuffer(this.mainType, buffer)
+    }
+
+    const self = this
+    let fullPacketData = null
+    function makeProxyHandler (path) {
+      return {
+        get: function (target, key, receiver) {
+          if (key === '_isProxy') return true
+          if (fullPacketData !== null) {
+            return Reflect.get(recursiveGet(fullPacketData, path, receiver), key, receiver)
+          }
+          if (!(key in target)) {
+            fullPacketData = self.proto.parsePacketBuffer(self.mainType, buffer).data
+            return Reflect.get(recursiveGet(fullPacketData, path, receiver), key, receiver)
+          }
+
+          let prop = Reflect.get(...arguments)
+          if (typeof prop === 'object' && !prop._isProxy) {
+            prop = new Proxy(prop, makeProxyHandler([...path, key]))
+            Reflect.set(target, key, prop, receiver)
+          }
+          return prop
+        }
+      }
+    }
```

Comment:
Avoid function creating - if 1000 "lazy" packets are processed, GC will have to track 1000 functions, even if they sharing the same code. Use `Function.bind` or tiny arrow functions.

Reply:
Thanks for the feedback! 
I'm under the impression that .bind is often slower than a closure? Either way, I believe a new function gets created. We should definitely try to benchmark a couple different solutions and pick the one that performs best empirically.

I pushed 39040d3 which uses .bind instead, does that look more reasonable to you?